OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,454 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_VULKAN_LAYER_LAYER_LOGIC_H_
+#define ORBIT_VULKAN_LAYER_LAYER_LOGIC_H_
+
+#include "OrbitBase/Logging.h"
+#include "VulkanLayerProducerImpl.h"
+#include "absl/base/casts.h"
+#include "vulkan/vk_layer.h"
+#include "vulkan/vulkan.h"
+
+namespace orbit_vulkan_layer {
+
+/**
+ * This class controls the logic of this layer. For the instrumented vulkan functions,
+ * it provides an `On*` function (e.g. for `vkQueueSubmit` there is `OnQueueSubmit`) that delegates
+ * to the driver/next layer (see `DispatchTable`) and calls the required functions for this layer to
+ * function properly. So it ties together the classes like the `SubmissionTracker` or the
+ * `TimerQueryPool`.
+ * In particular it performs the bootstrapping code (OnCreateInstance/Device) and the enumerations
+ * required by every vulkan layer.
+ *
+ * Usage: For an instrumented vulkan function "X" a common pattern from the layers entry (Main.cpp)
+ * `OnX` needs to be called to the controller.
+ *
+ * Note, the main reason, to not expose the vulkan functions directly in this class, is that this
+ * allows us to write tests that check if we glue the code correctly together and does the proper
+ * bootstrapping.
+ */
+template <class DispatchTable, class QueueManager, class DeviceManager, class TimerQueryPool,
+          class SubmissionTracker>
+class VulkanLayerController {
+ public:
+  // layer metadata
+  static constexpr const char* const kLayerName = "ORBIT_VK_LAYER";
+  static constexpr const char* const kLayerDescription =
+      "Provides GPU insights for the Orbit Profiler";
+
+  static constexpr const uint32_t kLayerImplVersion = 1;
+  static constexpr const uint32_t kLayerSpecVersion = VK_API_VERSION_1_1;
```

Comment:
So these need to be in sync with the json? I would at least add a comment about that.